09-intermediate-rails/oauth.md

```diff
@@ -173,95 +294,187 @@ class SessionsController < ApplicationController
     if user
       # User was found in the database
       flash[:success] = "Logged in as returning user #{user.name}"
-
     else
       # User doesn't match anything in the DB
       # Attempt to create a new user
       user = User.build_from_github(auth_hash)
 
       if user.save
         flash[:success] = "Logged in as new user #{user.name}"
-
       else
         # Couldn't save the user for some reason. If we
         # hit this it probably means there's a bug with the
         # way we've configured GitHub. Our strategy will
         # be to display error messages to make future
         # debugging easier.
         flash[:error] = "Could not create new user account: #{user.errors.messages}"
-        redirect_to root_path
-        return
+        return redirect_to root_path
       end
     end
 
     # If we get here, we have a valid user instance
     session[:user_id] = user.id
-    redirect_to root_path
+    return redirect_to root_path
   end
-end
+# ...
 ```
 
-## Wait!  How Do I Log Out!
-
-We have authentication working now, but the user cannot actually log out.  The user is instead permanently logged in.  To log the person out we need to clear the session.
+#### Our Helper Method `User.build_from_github`
 
-To start with we can add a route for logging out.
+What was the line `user = User.build_from_github(auth_hash)` above about? Observe our own implementation of a helper method to make the User from the `auth_hash`. Add this code so that your `User` model resembles this! Feel free to adjust this code with whatever makes sense on your projects.
 
 ```ruby
-delete "/logout", to: "sessions#destroy", as: "logout"
-```
-Then in the `SessionsController` we can add a method, `destroy` to handle logout.  
-
-```ruby
-#  app/controllers/sessions_controller.rb
-class SessionsController < ApplicationController
-  ...
-  def destroy
-    session[:user_id] = nil
-    flash[:success] = "Successfully logged out!"
-
-    redirect_to root_path
+class User < ApplicationRecord
+  def self.build_from_github(auth_hash)
+    user = User.new
+    user.uid = auth_hash[:uid]
+    user.provider = "github"
+    user.name = auth_hash["info"]["name"]
+    user.email = auth_hash["info"]["email"]
+
+    # Note that the user has not been saved.
+    # We'll choose to do the saving outside of this method
+    return user
   end
 end
 ```
 
-Lastly in our `/app/views/layouts/application.html.erb` file we can add links to log in and out.
+## 5. Wait! How Do I Log Out?!
 
-```erb
-<% if session[:user_id] %>
-  <%= link_to "Log out", logout_path, method: "delete"   %>
-<% else %>
-  <%= link_to "Login with Github", "/auth/github" %>
-<% end %>
-```
+We have authentication working now, but the user cannot actually log out.  The user is instead permanently logged in.  To log the person out we need to clear the session.
+
+1. Add a route for logging out:
+    ```ruby
+    delete "/logout", to: "users#destroy", as: "logout"
```

Comment:
Small thing, in the sessions lesson we used `:post` for logout, now we're using `:delete`... I prefer `delete`, but still.